### PR TITLE
#WIP specs for shifts

### PR DESCRIPTION
**Pull request by katafira**
@sergio-ocon @alexvkcr @delaluzparra
We do not know if the regex for dates should be in the model or in the tests. Right now it's in the tests.
Some pending tests were covered by the ones we developed:
  #pending "start_time is defined as HH:MM"
  #pending "start_time is between 00:00 and 23:59"
  #pending "end_time is defined as HH:MM"
  #pending "end_time is between 00:00 and 23:59"

### CODE REVIEW

**Comment by sergio-ocon** [comment]
The regex should be in the model, right now there is nothing there and the hour is only a text.
In fact, you should follow TDD.
Code it in the specs
See that it fails
Correct it in the code (with a quick fix) and see that it goes green
Refactor to a generic solution.
In that way, you know that
1 - Fails when it has to fail (no false positives)
2 - It is green when it has to be green (no false negatives)

File: .gitignore [hunk 1]
```diff
@@ -16,3 +16,6 @@
 !/log/.keep
 /tmp
 /.byebug_history
+
+# Ignore spec_helper.rb 
+/spec/spec_helper.rb
```
Comment (sergio-ocon) [inline .gitignore]: We don't ignore any other specs.

If you want to do this, it should be its own PR, mixin it with other PR does not make any sense

In addition, you can call rspec with your own configuration

And you can configure this in 
Global: ~/.rspec (i.e. in the user's home directory)
https://relishapp.com/rspec/rspec-core/v/3-4/docs/configuration/read-command-line-configuration-options-from-files

File: spec/spec_helper.rb [hunk 2]
```diff
@@ -89,4 +89,11 @@
   # as the one that triggered the failure.
   Kernel.srand config.seed
 =end
+  config.color = true
+
+  # Use color not only in STDOUT but also in pagers and files
+  config.tty = true
+
+  # Use the specified formatter
+  config.formatter = :documentation # :progress, :html, :textmate
```
Comment (sergio-ocon) [inline spec/spec_helper.rb]: Delete this

File: spec/models/shift_spec.rb [hunk 3]
```diff
@@ -48,12 +49,28 @@
...
+  end
+
+  # DONE pending "prebooked is lower than the capacity of the room it belongs to"
   pending "relations are nullified whtn the shift is detroyed"
   pending "it gives back the number of spaces available"
-  pending "start_time is defined as HH:MM"
-  pending "start_time is between 00:00 and 23:59"
-  pending "end_time is defined as HH:MM"
+  #pending "start_time is defined as HH:MM"
+  #pending "start_time is between 00:00 and 23:59"
```
Comment (sergio-ocon) [inline spec/models/shift_spec.rb]: Why are you commenting out this pending?
You can delete it, but I don't understand commenting it.

File: spec/models/shift_spec.rb [hunk 4]
```diff
@@ -28,17 +29,17 @@
...
   end
 
   it "end_time is a time" do
-    shift.end_time="lunes"
-    shift.valid?
-    expect(shift).not_to be_valid
-    expect(shift.errors[:end_time]).to include "no puede estar vacío"
+    expect(shift.end_time).to be_instance_of(String)
+    #Should the regular expression be defined in the model?
+    regex = '^\d{0,4}-(0[123456789]|1[012])-(0[123456789]|[12]\d|3[01])\s([01]\d|2[0123]):[012345]\d:[012345]\d$'
```
Comment (sergio-ocon) [inline spec/models/shift_spec.rb]: Yes, the regular expression should be part of the validation of the model, you need to add that code to the model as part of doing this SPEC.

TDD

File: app/models/shift.rb [hunk 5]
```diff
@@ -16,6 +17,15 @@ def sites_available
   end
 
   private
+    #Custom validation to ensure sites_available greater than prebooked
+    def sites_available_gt_0
+      if sites_available?
+        return true
+      else
+        errors.add(:prebooked, 'More prebooked than sites_available')
```
Comment (sergio-ocon) [inline app/models/shift.rb]: Please add the errors in Spanish or add them as i18t.n, so they can be used in both languages

**Comment by alexvkcr** [comment]
Branch rebased to match original repository

File: spec/models/shift_spec.rb (hunk 3 above)
Comment (alexvkcr) [inline spec/models/shift_spec.rb]: We left it commented to ensure you agreed to delete them. Deleting them in next commits

**Comment by alexvkcr** [comment]
I updated master and rebased through the shell and uploaded the branch updated.

After that, GitHub still said that it was not synchronized with branch master at 'sergio-ocon/lastablas' and it suggested to click a bottom to perform that action.

Now, it looks like in the commit history of this branch, there are duplicated commits(but with different hash). What should we do about this?

**Comment by sergio-ocon** [comment]
I've updated some lines of code, and fixed most of the problems, you need to rebase again.

Now travis should be green

File: app/models/shift.rb [hunk 6]
```diff
@@ -29,6 +31,17 @@ def sites_available_gt_0
       end
     end
 
+    def regexed_time
+      regex = /^\d{0,4}-(0[123456789]|1[012])-(0[123456789]|[12]\d|3[01])\s([01]\d|2[0123]):[012345]\d:[012345]\d$/
```
Comment (sergio-ocon) [inline app/models/shift.rb]: This does not work
http://rubular.com/
And try it...

Comment (sergio-ocon) [inline app/models/shift.rb]: ^(\d|([0-1]\d)|2[0-3]):[0-5][0-9]$

Try this for all cases, make an rspec for this. Test the function.
1:12
0:01
8:12
08:12
25:12 # debe fallar
01:99 #debe fallar
24:01 # debe fallar
etc

**Comment by sergio-ocon** [comment]
Please rebase this before merging

File: .byebug_history [hunk 7]
```diff
@@ -1,4 +1,63 @@
 continue
+shift.errors
```
Comment (sergio-ocon) [inline .byebug_history]: I don't know why .byebug_history is included, it is in my .gitignore file. 
Perhaps a new PR?

File: app/models/shift.rb [hunk 8]
```diff
@@ -19,6 +21,30 @@ def sites_available
...
+      if sites_available?
+        return true
+      else
+        errors.add(:prebooked, 'More prebooked than sites_available')
+        return false
+      end
+    end
+
+    def regexed_time
+      regex = /^\d{0,4}-(0[123456789]|1[012])-(0[123456789]|[12]\d|3[01])\s([01]\d|2[0123]):[012345]\d:[012345]\d$/
```
Comment (sergio-ocon) [inline app/models/shift.rb]: You can use the regex I provided before or look here:
http://stackoverflow.com/questions/7536755/regular-expression-for-matching-hhmm-time-format
That one is not working at all
http://rubular.com/

**Comment by alexvkcr** [comment]
Working right now to squash some commits for simplicity and make a new PR

File: app/models/shift.rb (hunk 8 above)
Comment (alexvkcr) [inline app/models/shift.rb]: It looks like now it's passing the tests.

**Comment by alexvkcr** [comment]
Kata is tasked with closing this PR, this line of work moves to #8
